client/js/widgets/scoreboard.js

```diff
@@ -0,0 +1,317 @@
+    for (let i=0; i < values.length; i++)
+      $a(cellType, tr)[i].innerText = values[i];
+    return tr;
+  }
+
+  // Compute desired text color from luminance of background
+  // See https://stackoverflow.com/questions/3942878
+  getFgColor(bgColor) {
+    function getHexColor(color) {
+      const a = document.createElement('div');
+      a.style.color = color;
+      const colorStr = window.getComputedStyle( document.body.appendChild(a) ).color;
+      const colors = colorStr.match(/\d+/g).map(function(a){ return parseInt(a,10); });
+      document.body.removeChild(a);
+      return colors;
+    }
+    const bg = getHexColor(bgColor);
+    return bg[0]*0.299 + bg[1]*0.587 + bg[2]*0.114 > 186 ? "#000000" : "#ffffff";
+  }
+
+  tableCreate(seats) {
+    /* This routine creates the HTML table for display in the scoreboard. It is
+     * complicated by the fact that the `seats` property can be either an array of
+     * seat IDs or an object whose keys are team names and each of whose values is an
+     * array of seat IDs.
+     * There are two major sections: the first computes the pScores array, which contains
+     * the array of scores, either for seats or for teams. The second section uses the
+     * pScores array to construct the HTML table.
+     * There are lots of other things going on, to get the totals line, round names, etc
+     * correct.
+     */
+
+    // First deal with cases where we just need to clear the table; invalid set of seats.
+    // We choose here to regard a result of [] or {} as a valid set of seats/teams with no entries.
+    if(seats===null) {
+      if(this.tableDOM)
+        this.tableDOM.innerHTML = '';
+      return
+    }
+      
+    const showTotals = this.get('showTotals');
+    const scoreProperty = this.get('scoreProperty');
+    let sortField = this.get('sortField');
+
+    // Compute number of scoring rounds to show and create round names table
+    let rounds = this.get('rounds'); // User-supplied round names
+    let numRounds=0;
+    const arrayOfSeats = Array.isArray(seats) ? seats : Object.keys(seats).reduce((union,key) => union.concat(seats[key]), []);
+    for (let i=0; i < arrayOfSeats.length; i++) {
+      const score = arrayOfSeats[i].get(scoreProperty);
+      if(Array.isArray(score) && score.length > numRounds)
+        numRounds = score.length;
+    }
+    if(this.get('showAllRounds') && Array.isArray(rounds))
+      numRounds = Math.max(rounds.length, numRounds);
+
+    if(Array.isArray(rounds))
+      rounds = rounds.concat(Array(numRounds).fill('')).slice(0,numRounds);
+    else
+      rounds = [...Array(numRounds).keys()].map(i => i+1);
+    if(showTotals)
+      rounds.push('Totals');
+    rounds.unshift(this.get('roundLabel'));
+
+    // Fill scores array. pScores[i][0] is player name or team name, last is total
+    // (or last score if showTotals is false)    
+    let pScores = []; 
+    if(Array.isArray(seats)) { // Show individual seats
```

Comment:
I think this could be refactored and simplified so that teams or seats work mostly the same. I.e. you could treat both as if they are teams where the seats case is just autogenerated teams of a single player per seat.

Reply:
Could be done, yes. But I'm not sure it's worth it. There is lots of code in the seats case that doesn't apply to teams (sorting, player name colors). So at some point you'd have to differentiate, and in the end I'm not sure the code would be shorter or clearer.